api: add base contract wrappers

ixje opened the pull request:
fix #149

This PR is about adding the base code for calling smart contracts conveniently over RPC. From the docstrings of the `wrappers` module

```
Convenience wrappers for calling smart contracts via RPC.
* The most specific wrappers in this module are for NEOs native contracts
  * NeoToken, GasToken TODO: add remaining contracts
* One step up are wrappers for contracts following the NEP-17 Token standard (`NEP17Contract`) and NEP-11 NFT standard
(`NEP11Contract`)
* As last resort there is the `GenericContract` which can be used for calling arbitrary functions on arbitrary contracts

Obtaining the results of the functions on the wrapped contracts is done through one of 3 methods
1. test_invoke() - does not persist to chain. Costs no gas.
2. invoke() - does persist to chain. Requires signing and costs gas.
3. estimate_gas - does not persist to chain. Costs no gas.
```

### Current state

* ~~The `NEP11Contract` generic contract is not yet implemented.~~
* `invoke()` is not yet implemented
* The `Config` isn't thought out yet. Just contains some minimal stuff now

### Feedback
This is a work in progress. I'm looking for feedback on the API in terms of
* is easy is it to use it (considering there is not yet any documentation other than inline)?
* how pythonic does it feel?
* what do you like/dislike about how it works?
* what can be improved?

### Example usage

Here are 3 examples to get your started

-edit- code changed. New example code can be found in this comment https://github.com/CityOfZion/neo-mamba/pull/184#issuecomment-1265176954

ixje commented:
Other questions I have and like to get input on

1. Should we support NeoToken "set" functions that only work if you're part of the consensus committee? e.g. `SetGasPerBlock`
  yes, means complete API support
  no, means we don't show functions that 99% of the users can't use anyway making the function list a bit cleaner
2. Should we support NeoToken "GetNextBlockValidators"
  yes, means more complete support of all public methods available on the contract
  no, means hiding a function that probably nobody is going to use. It's still available for calling through other means like the RPCClient or by manual extension
3. The current design has 3 "free" functions for calling contract methods/scripts
   1. test_invoke()
   2. invoke()
   3. estimate_gas()

All of them require accepting some configuration (what RPC host to use, what account to use as sender, what additional signers to add?) for accurate results and in case of `invoke` to work at all. If we would bind these functions to a single object (think `NetworkFacade` in `neon-js`) then we could skip passing this config into the "free" functions every time e.g.

```python
facade = Facade(config)
await facade.test_invoke(f)
await facade.invoke(f, <signing func>)
await facade.estimate_gas(f)
```
Do we like this more than the current way?

4.  `ContractMethodFuture` vs `ContractScriptFuture` vs ...?
 Currently the return type of the functions on the wrapped smart contracts are of type `ContractMethodFuture`. This implies a single method call. While True in most cases it does not hold in all. Also functions like `tes_tinvoke()` take this as input, while in reality it can contain a script that does much more (especially when using the `GenericContract`). I'm having a hard time to decide whether to keep calling it `ContractMethodFuture` which suggests a future result (where `test_invoke()` will get that result) for a single call (which reads nicely if you're reading the method signatures on e.g. `NeoToken`). On the other hand it is misleading when reading the signature of `test_invoke()` because the following is equally valid for more advanced cases
```python
 sb = ScriptBuilder()
 sb.emit_contract_call(some_hash, "func1")
 sb.emit_contract_call(some_hash, "func2")
 sb.emit_contract_call(some_hash, "func3")
 f = future_contract_method_result(sb.to_array())  # this method should then also rename to something like `future_script_result`
 testinvoke(f)
```

lock9 commented:
Hi @ixje ,



> 1. Should we support NeoToken "set" functions that only work if you're part of the consensus committee? e.g. `SetGasPerBlock`
>    yes, means complete API support
>    no, means we don't show functions that 99% of the users can't use anyway making the function list a bit cleaner

I would like to propose a different approach:  
Could we make mamba 'agnostic' (use example 2), and have the other interfaces part of Props?
This would allow more flexibility in updating these templates. I think that for "Neo", the answer may be easier (add the extra methods or not), but what about other contracts? I feel that it won't scale well if we add everything to mamba. 


> The current design has 3 "free" functions for calling contract methods/scripts
> 
> 1. test_invoke()
> 2. invoke()
> 3. estimate_gas()
> 
> All of them require accepting some configuration (what RPC host to use, what account to use as sender, what additional signers to add?) for accurate results and in case of `invoke` to work at all. If we would bind these functions to a single object (think `NetworkFacade` in `neon-js`) then we could skip passing this config into the "free" functions every time e.g.
> 
> ```python
> facade = Facade(config)
> await facade.test_invoke(f)
> await facade.invoke(f, <signing func>)
> await facade.estimate_gas(f)
> ```
> 
> Do we like this more than the current way?


I don't think this will solve the problem. I think we need something more similar to the IconDapp TS example:
https://github.com/CityOfZion/neo3-invoker/blob/3660186c1a44a766cd6c132e15a11035bd7bc96c/packages/neon-invoker/src/index.ts#L33

It would be great to have some sort of unified interface for both of them.  This will make documentation and maintenance easier. They don't have to be identical, but we must have some minimum standards for the names and expected behavior

ixje commented:
> Hi @ixje ,
> 
> > 1. Should we support NeoToken "set" functions that only work if you're part of the consensus committee? e.g. `SetGasPerBlock`
> >    yes, means complete API support
> >    no, means we don't show functions that 99% of the users can't use anyway making the function list a bit cleaner
> 
> I would like to propose a different approach: Could we make mamba 'agnostic' (use example 2), and have the other interfaces part of Props? This would allow more flexibility in updating these templates. I think that for "Neo", the answer may be easier (add the extra methods or not), but what about other contracts? I feel that it won't scale well if we add everything to mamba.

I think there might be a misunderstanding. The 3 examples given are not mutually exclusive, they are all valid. We have dedicated wrappers for NEOs native contracts because they almost never change. Then we have `NEP17Contract` and `NEP11Contract` which you can use as a base class for talking to any NEP17 or NEP11 token. The `NeoToken` class is actually build on top of `NEP17Contract`. Having said that example 2 could as well have a looked like this
```python
lyrebird_token = NEP17Contract(types.UInt160.from_string("0xaa38358290234a181d6dd60baa7aa921822dbb21")
print(await test_invoke(lyrebird_token.symbol(), config)
```

---------

> 
> > The current design has 3 "free" functions for calling contract methods/scripts
> > 
> > 1. test_invoke()
> > 2. invoke()
> > 3. estimate_gas()
> > 
> > All of them require accepting some configuration (what RPC host to use, what account to use as sender, what additional signers to add?) for accurate results and in case of `invoke` to work at all. If we would bind these functions to a single object (think `NetworkFacade` in `neon-js`) then we could skip passing this config into the "free" functions every time e.g.
> > ```python
> > facade = Facade(config)
> > await facade.test_invoke(f)
> > await facade.invoke(f, <signing func>)
> > await facade.estimate_gas(f)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do we like this more than the current way?
> 
> I don't think this will solve the problem. I think we need something more similar to the IconDapp TS example: https://github.com/CityOfZion/neo3-invoker/blob/3660186c1a44a766cd6c132e15a11035bd7bc96c/packages/neon-invoker/src/index.ts#L33

Please elaborate what I'm supposed to see in the IconDapp TS example? What I see is similarly a facade class (`NeonInvoker`) with a `testInvoke` function that internally uses a RPC config and account that have been supplied during creation of the facade class.


> It would be great to have some sort of unified interface **for both of them**. This will make documentation and maintenance easier. They don't have to be identical, but we must have some minimum standards for the names and expected behavior

what are you referring to with "both of them"?

lock9 commented:
> Please elaborate what I'm supposed to see in the IconDapp TS example? What I see is similarly a facade class (`NeonInvoker`) with a `testInvoke` function that internally uses a RPC config and account that have been supplied during creation of the facade class.

Two things:

First: When we concatenate multiple scripts, we may want to add an 'abort if fails' between each invocation. This is a must for advanced users:
![image](https://user-images.githubusercontent.com/38396062/191573344-76f92bc6-2e6b-4ebf-afff-72f1d472b48e.png)

Second:
There is something 'odd' for me. I don't know if we should send the config all the time. I feel that you would do something like 'blockchain.test_invoke(params)', and not 'test_invoke(blockchain, params). 

I think I'm confused about all the options. The ```facade``` is close to what I expected, but at the same time, it's not the same thing.


>what are you referring to with "both of them"?

The N3-Invoker and the Python interfaces. Note: N3 invoker are the helper classes / features we extracted from Icon Dapp. 

I'll ping you tomorrow to discuss this

ixje commented:
> First: When we concatenate multiple scripts, we may want to add an 'abort if fails' between each invocation. This is a must for advanced users: ![image](https://user-images.githubusercontent.com/38396062/191573344-76f92bc6-2e6b-4ebf-afff-72f1d472b48e.png)

This feature exists on `NEP17Contract.transfer_multi()` but I think it is incorrect to have this in `testInvoke` because this will only work as expected when the function you're calling returns a `bool` (like `transfer`) but not in other cases.

> Second: There is something 'odd' for me. I don't know if we should send the config all the time. I feel that you would do something like 'blockchain.test_invoke(params)', and not 'test_invoke(blockchain, params).

It depends, one scenario could be; create 1 config which uses the company shared account and call user level specific functions with it. Then have another config for admins only which I pass to `invoke()` whenever I want to do administrative functions. However, the same can be done with multiple facades.

> what are you referring to with "both of them"?
>>The N3-Invoker and the Python interfaces. Note: N3 invoker are the helper classes / features we extracted from Icon Dapp.

I don't think we should have 2 different projects/languages try to keep the same interface. That's just going to increase the complexity for no reason

ixje commented:
Another topic I'd like input on.

**address vs script hash as arguments**
These wrappers are the most user-end facing part of the code base. While those reading this PR might know that a NEO wallet address is nothing more than a script hash (stored in a `UInt160`) with some extra (`base58check`) encoding , many newer users might not know this.  If we look at for example the `NEP-17` `transfer()` method, that now takes a UInt160 as `source account`, a UInt160 as `destination account` and an `amount`. End-users most likely end up using it like this 

```python
source_account = utils.address_to_script_hash("NV96QgerjXNmu4jLdMW4ZWkhySVMYX52Ex")
destination_account = utils.address_to_script_hash("NeT85LegV3eNQCJjLjhUSgsTxuXcfra2bp")

token.transfer(source_account, destination_account, 10)
```
We still have the option to instead make the parameters NEO addresses so we get
```python
token.transfer("NV96QgerjXNmu4jLdMW4ZWkhySVMYX52Ex", "NeT85LegV3eNQCJjLjhUSgsTxuXcfra2bp", 10)
```

This would apply to the following functions
* NEP17Contract
   * balanceOf
   * transfer
* NEP11Contract
   * balanceOf
   * tokensOf
   * transfer
   * ownerOf (the return value)
* NeoToken
   * get_unclaimed_gas 
   * candidate_vote
* PolicyContract
   * is_blocked(account)
   * block_account
   * unblock_account
  
Ideas? Good/bad?

ixje commented:
After some internal discussions the following has been decided and changed
* We will not add support for the `setXXX` functions on NeoToken that require one to be part of the consensus committee (question 1. of https://github.com/CityOfZion/neo-mamba/pull/184#issuecomment-1253703580)
* The design changed a little bit to move the free functions (test_invoke,invoke, estimate_gas) to be members of `ChainFacade` class (question 3. of https://github.com/CityOfZion/neo-mamba/pull/184#issuecomment-1253703580)

    This eliminates the need for passing a `Config` to every free function call. Calling the same examples as before in the new way looks like this

    ```python
    from __future__ import annotations
    import asyncio
    from neo3.core import types
    from neo3.api import unwrap
    from neo3.api.wrappers import NeoToken, ChainFacade, GenericContract, NEP17Contract
    from neo3.network.payloads import verification
    from neo3.wallet import utils as utils


    async def example1():
        # Example 1 - call a function on the NeoToken wrapper
        facade = ChainFacade.node_provider_testnet()
        neo = NeoToken()
        res = await facade.test_invoke(neo.candidates_registered())

        for candidate in res:
            print(f"{candidate.address} {candidate.votes:>8}")


    async def example2():
        # Example 2 - call a function on an arbitrary NEP-17 contract, we'll use GhostMarket as example
        facade = ChainFacade.node_provider_mainnet()
        ghostmarket_script_hash = types.UInt160.from_string("0x9b049f1283515eef1d3f6ac610e1595ed25ca3e9")
        gm_token = NEP17Contract(ghostmarket_script_hash)

        source_account = utils.address_to_script_hash("NhWVzumNarG7x41c8csTor9UvocoSUrQQL")
        destination_account = utils.address_to_script_hash("NeT85LegV3eNQCJjLjhUSgsTxuXcfra2bp")
        signer = verification.Signer(source_account, verification.WitnessScope.CALLED_BY_ENTRY)

        res = await facade.test_invoke(
            gm_token.transfer(source_account, destination_account, 1),
            signers=[signer]
        )
        print(f"Transferred? {res}")


    async def example3():
        # Example 3 - call arbitrary functions on an arbitrary contract
        facade = ChainFacade.node_provider_testnet()
        puppet_hash = types.UInt160.from_string("0x76a8f8a7a901b29a33013b469949f4b08db15756")
        puppet = GenericContract(puppet_hash)
        response = await facade.test_invoke(puppet.call_function("total_accounts"))
        print(f"Puppet accounts: {unwrap.as_int(response)}")

    if __name__ == "__main__":
        asyncio.run(example1())
        asyncio.run(example2())
        asyncio.run(example3())
    ```

* Added 2 additional functions `test_invoke_raw` and `invoke_raw`. The most high level wrappers like `NeoToken` will unwrap the contract call response into nice and easy to use results (by some argued as front-end results). However, this might not always be desired. The `_raw` versions will return the "raw" [ExecutionResult](https://github.com/CityOfZion/neo-mamba/blob/17a3816fe511b34fa317e38a0781a2e22a67b724/neo3/api/noderpc.py#L265) such that it can be processed as preferred

Stay tuned, more to come..

ixje commented:
The wrappers in this PR look as follows

![wrappers](https://user-images.githubusercontent.com/6625537/193556067-28032361-bc39-472f-aaf0-de9f6f79cd5c.png)

* Green are generic type wrappers 
* Blue are very specific
* White are just building blocks and should not be used on their own. They are shown to make the picture complete.

ixje commented:
I personally found the method names used in the NEP-11 standard to not be the best chosen. Combine this with the fact that they use method overloading (which is not supported in Python) and I had enough reason to not keep the names on the wrappers strictly the same as in the standard. 

These are the method names that I changed and would like to hear you opinion on this
| NEP-11 standard name | New name           | Description                |
|----------------------|--------------------|----------------------------|
| balanceOf            | total_owned_by     | total NFT count for owner  |
| tokensOf             | token_ids_owned_by | returns token ids          |
|                      |                    |                            |
| **Divisible NFT**    |                    |                            |
| ownerOf              | owners_of          | list of all partial owners |

The name `token_ids_owned_by` is based on the following reasoning; there are a couple of methods on the NEP11 classes that take a `token_id` as argument (`properties()`, `transfer()`, `owner_of()`, `balance_of()`). If you happen to call one of these methods and see that you need to obtain a `token_id` then going through all available methods on the class to find that `tokensOf()` is the method that gives you token id's didn't seem like a good experience to me. `token_ids_owned_by` tells in the name that it will return token ids.

ixje commented:
fyi; I plan to merge this by the end of the week because it is starting to block other PRs
* `invoke()` will be implemented in a separate PR because there's quite some stuff to discuss around it
* the remaining native contracts will be implemented in a separate PR

ixje commented:
Note: you can also call arbitrary scripts through the `ChainFacade` without having to create wrappers e.g.

```python
    facade = ChainFacade.node_provider_mainnet()
    script = vm.OpCode.PUSH1 + vm.OpCode.RET
    print(await facade.test_invoke(ContractMethodResult[int](script, unwrap.as_int)))
```

lock9 commented on neo3/api/unwrap.py:
```diff
@@ -0,0 +1,164 @@
+    check_state_ok(res)
+    if idx > len(res.stack) - 1:
+        raise ValueError("Too few result items")
+    return res.stack[idx]
```
I saw that this method `item `returns a `StackItem`, but the other methods use the `ExecutionResult`. How is the conversion done? It seems possible to run a transaction and get multiple results from it, but it's not clear to me how

ixje replied:
What other methods are we talking about? All functions in `unwrap` take `ExecutionResult`  as input. Internally they call a similarly named function on the `StackItem` that is returned.

Pseudo; this
```python
result: ExecutionResult = Object()  # assume this is a real result
stack_item = unwrap.item(result, 0)  # checks VM state is HALT and get the stack item at index 0 from the `ResultStack`
my_int = stack_item.as_int()

```
is the same as
```python
result: ExecutionResult = Object()
my_int = unwrap.as_int(result)
```

The implementation of the unwrap functions used to do a bit more. But I moved that to the `StackItem` class itself so you can also convert from there (in case you prefer to work with the `ResultStack` yourself)

lock9 reviewed (commented):
It looks good, but it doesn't have any examples.  
I suppose they are compatible, but just to illustrate, could you explain/exemplify how an airdrop would work? 
One thing that always bothers me is: I never know if we should calculate the fee or not. I don't know if the test invoke will be done automatically.

ixje commented:
> It looks good, but it doesn't have any examples. 

No examples in `/examples/` is intentional. Examples to me make sense when they can be performed on the real network, not with a `test_invoke`. As said in https://github.com/CityOfZion/neo-mamba/pull/184#issuecomment-1266815878 `invoke` is planned for a later PR. Also, documentation out side of the docstrings (meaning the ones in `/docs/`) need a major overhaul with this new mamba version. Something I'm postponing until the end as there's still quite a few PRs to come and discuss.

> I suppose they are compatible, but just to illustrate, could you explain/exemplify how an airdrop would work? One thing that always bothers me is: I never know if we should calculate the fee or not. I don't know if the test invoke will be done automatically.

You're slightly ahead of the scope of this PR, as fee calculation really only matters in the `invoke` case and as said earlier the details and implementation of `invoke` are for another PR. But how it would look is as follows

```python
    token = NEP17Contract(contract_hash)
    await facade.invoke(token.transfer_multi(source, addresses, 1))
```
Fee calculation will be done automatically with options to override or pad the fees.

EdgeDLT commented:
Been trying this out this weekend and it feels pretty intuitive to use. I made a basic version of an invocation utility I made with Neon.js and got [similar results](https://user-images.githubusercontent.com/21316801/194934844-0367d78b-f246-4662-984c-55a3e01897f4.mp4) relatively quickly. Insert an Xzibit meme about putting a wrapper on your wrapper here.

To weigh in on a few of the discussion points so far:

* I like the way config is handled in the updated version. As you mentioned @ixje, there's always the option to create more facades if I need to use a different account for some reason.
* Agree with the decision not to add committee methods. They can always be done with the generic interface if necessary (and probably not even then, a signing facility is all you realistically need unless you're the initiator of the committee tx).
* I think it would be good to support `Union[str, UInt160]` for addresses/script hashes where appropriate. I think the mental distinction of an address vs a script hash as representing a wallet vs a contract is helpful for beginners, this facilitates that a little bit. They can find out about the reality later.
* The `Unwrap` helpers are nice! Would it be beneficial to have `as_exception()` for catching errors and `as_any()` for catching when methods fail silently with no return value? Unfortunately there are many contracts that do this, including NeoNS. I understand if you think those are unnecessary though, since you can already just take `ExecutionResult.exception` or default to `None` when the return type doesn't match expectations.
* Agree with the method name choices for NEP-11. It should make it more accessible, as long as how they correlate to NEP-11's methods is documented and easy to locate. In fact, maybe it belongs in the docstrings so the correlations show up in Intellisense?
* What do you think about having more types? So for example when providing an integer argument, rather than using `int(parameter)` I might have a `types.Integer.from_string(parameter)` instead. This would be like `sc.ContractParam` in Neon.js. Advantage in my eyes is that I can easily see the options and pick the correct type I want and know it will take care of basic conversions. Maybe it's unnecessary though.

I haven't tried any custom scripts yet, but that's a cool feature that I'm eager to play around with.

Looking good 👍

ixje commented:
> Been trying this out this weekend and it feels pretty intuitive to use. I made a basic version of an invocation utility I made with Neon.js and got [similar results](https://user-images.githubusercontent.com/21316801/194934844-0367d78b-f246-4662-984c-55a3e01897f4.mp4) relatively quickly. Insert an Xzibit meme about putting a wrapper on your wrapper here.

That looks cool :hushed:
> 
> To weigh in on a few of the discussion points so far:
> 
> * The `Unwrap` helpers are nice! Would it be beneficial to have `as_exception()` for catching errors and `as_any()` for catching when methods fail silently with no return value? Unfortunately there are many contracts that do this, including NeoNS. I understand if you think those are unnecessary though, since you can already just take `ExecutionResult.exception` or default to `None` when the return type doesn't match expectations.

I think that in general the user has the base assumption that calling a smart contract method will succeed. Now reality can be different, for which there is a [check_state_ok()](https://github.com/CityOfZion/neo-mamba/blob/18d2afd5c42c381fed936309c77e8bc6dcaa43b1/neo3/api/unwrap.py#L23) (which is arguably similar to `as_exception`) and is internally called for all `as_<type>` functions. So if the user assumes (e.g. based on the manifest) that a function returns a `str` and passes `unwrap.as_str` as post processor but the result returns an exception, then `check_state_ok()` will fail and raise an error. On top of that all `as_<type>` functions also check if the result is empty or of a different type than requested and raise an error. That means you should actually call contract methods with
```python
try:
   results = await facade.test_invoke(neo.symbol())
except ValueError as e:
   # something failed
   print(f"invoke failed with: {e}")
```
which matches with Python's preferred "Ask for forgiveness, not permission" coding style. Perhaps if you can give me 1 or 2 examples I can better judge if an `as_exception` or similar helper makes sense to add after all. Still open to the idea, just not convinced yet on the usefulness.

As for `as_any`. I think the request here is kind of like `as_optional_int`, `as_optional_str` etc. These also get caught with the same `try/except` above. The question is probably, do you want to be able to differentiate between an exception that is caused by a `FAULT` state of the virtual machine (which can be many reasons) and a "silent exception" by the smart contract that "fails" for some internal reason? Again I think that users assume a function succeeds and any failure means abort and reconsider what to do. Happy to hear a case where differentiating is important so I have a bit of context to think about it some more.

> * Agree with the method name choices for NEP-11. It should make it more accessible, as long as how they correlate to NEP-11's methods is documented and easy to locate. In fact, maybe it belongs in the docstrings so the correlations show up in Intellisense?

Good point on the docstrings. I initially planned to have this NEP-11 mapping in some external documentation, but should indeed also add this to docstrings for those who don't read external docs.

> * What do you think about having more types? So for example when providing an integer argument, rather than using `int(parameter)` I might have a `types.Integer.from_string(parameter)` instead. This would be like `sc.ContractParam` in Neon.js. Advantage in my eyes is that I can easily see the options and pick the correct type I want and know it will take care of basic conversions. Maybe it's unnecessary though.

I think it's unnecessary. Allow me to elaborate why. The big different between `neon-js` is that almost everything in `neon-js` internally is either a `number` or a `string` where in mamba everything is a specific type (and most of them are even native types). 

Here's the comparison

| Contract Type             | Neon-js                | Mamba                  |
| ---------------- | ---------------------- | ---------------------- |
| Any              | string \| null          | All of the below       |
| Boolean          | string \| number \| bool | bool                   |
| Integer          | string \| number        | int \| types.BigInteger |
| ByteArray        | string                 | bytes \| Iserializable  |
| String           | string                 | str                    |
| Hash160          | string                 | types.UInt160          |
| Hash256          | string                 | types.UInt256          |
| PublicKey        | string                 | cryptography.ECPoint   |
| Signature        | not supported          | bytes                  |
| Array            | \[\]                   | list                   |
| Map              | not supported          | dict                   |
| InteropInterface | not applicable when called from outside the VM         | N/A          |
| Void             | special case           | None                   |
 
This means that with mamba you can just work with the normal Python types or those that are already used everywhere in the SDK and not have to worry about byte ordering, encoding etc. Note that this statement holds not just for inputting arguments in the wrappers of this PR, but also for all `NeoRpcClient` methods ([invoke_function](https://github.com/CityOfZion/neo-mamba/blob/18d2afd5c42c381fed936309c77e8bc6dcaa43b1/neo3/api/noderpc.py#L964) in particular) and for any custom script you build using `ScriptBuilder` ([ref](https://github.com/CityOfZion/neo-mamba/blob/18d2afd5c42c381fed936309c77e8bc6dcaa43b1/neo3/vm.py#L262))

The `sc.ContractParam` approach imo is just a bandage around the lack of types used in `neon-js`. As for your `types.Integer.from_string(parameter)` this is built into python as `int("some number as string")` and then it just works again. 

> I haven't tried any custom scripts yet, but that's a cool feature that I'm eager to play around with.
> 
> Looking good +1

:raised_hands: thanks for the input!

EdgeDLT commented:
Thanks for all your thoughts @ixje, agree with all your points.

> So if the user assumes (e.g. based on the manifest) that a function returns a str and passes unwrap.as_str as post processor but the result returns an exception, then check_state_ok() will fail and raise an error.

The problem I was having here (the "failing silently" issue) was that some methods don't FAULT when they fail. The one I found that did this was NeoNS `getRecord`. 

The happy path returns a string value as expected. Providing a domain argument that doesn't exist will FAULT as expected. But providing a correct domain but with an incorrect record type results in HALT and returns `Any` with value undefined. So the `check_state_ok()` step was missing this part.

That said, I refactored the code to follow your example, leaving it to the `as_str()` to throw a `ValueError`, which works as expected. Seems you have all the bases covered already, I just need to learn to write more Pythonic code :smile:
